alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs

```diff
@@ -58,15 +63,22 @@ instance FromCBOR (Annotator PlutusData) where
 -- The newtype will memoize the serialized bytes. The strategy is to replace
 -- PlutusData  with the correct type
 
-newtype Data era = DataConstr (MemoBytes (PlutusData))
+newtype Data era = DataConstr (MemoBytes PlutusData)
   deriving (Eq, Ord, Generic, ToCBOR, Show)
+  deriving newtype (SafeToHash)
 
 deriving via
   (Mem PlutusData)
   instance
     (Era era) =>
     FromCBOR (Annotator (Data era))
 
+-- If and when the 'era' in (Data 'era') is not just a phantom parameter
+-- we will have to change this to a HashAnnotated instance, because then
+-- 'era' will determine 'crypto'
```

Comment:
As we discovered this is solved by using UndecideableInstances

If we write what we want to:
instance (Crypto era ~ c) => HashAnnotated (Data era) EraIndependentData c

we get an error message (Data era) does not determine (c), which of course it does not. But what
the type inference is not paying attention to is the constraint (Crypto era ~ c), by substituting (Crypto era) for c,  It is now obvious that  (Data era) determines (Crypto era), since Crypto is a function of era.  Adding undecideabe instances, allows the type checker to run the simplifier, where this substitution becomes possible, and everything works.